src/ChunkMap.cpp

```diff
@@ -2695,44 +2696,87 @@ void cChunkMap::TickBlock(int a_BlockX, int a_BlockY, int a_BlockZ)
 
 
 
-void cChunkMap::UnloadUnusedChunks(void)
+size_t cChunkMap::UnloadUnusedChunks(size_t a_Amount)
 {
+	size_t UnloadedTotal = 0;
+	bool UnloadAll;
+
+	if (a_Amount == 0)
+	{
+		UnloadAll = true;
+		a_Amount = m_World->GetNumChunks();
+	}
+	else
+	{
+		UnloadAll = false;
+	}
+
+	if (a_Amount == 0)
+	{
+		return 0;
+	}
+
 	cCSLock Lock(m_CSLayers);
-	for (auto & itr: m_Layers)
+	if (UnloadAll)
 	{
-		cChunkLayer & Layer = itr.second;
-		Layer.UnloadUnusedChunks();
-	}  // for itr - m_Layers
+		size_t Amount = 0;  // Just for debugging
+		for (auto & itr : m_Layers)
+		{
+			cChunkLayer & Layer = itr.second;
+			Amount += Layer.UnloadUnusedChunks(a_Amount);
+		}
+		ASSERT(Amount == a_Amount);
+		return a_Amount;
+	}
+	else
+	{
+		cFastRandom Rand;
+		// Choose 10 layers randomly and unload them.
+		// Stop if we've unloaded enough.
+		for (int i = 0; i < 10; ++i)
+		{
+			int ChosenLayerNum = Rand.NextInt(static_cast<int>(m_Layers.size()));
+			auto itr = m_Layers.begin();
+			while (ChosenLayerNum > 0)
```

Comment:
That's still `O(n)` isn't it? I am sort of disappointed that `std::map` has no built in way to pick a random element in `O(log n)`. That should be technically possible with a binary search tree.